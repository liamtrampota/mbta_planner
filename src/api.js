const MTA_API_URL = "https://api-v3.mbta.com";

const options = {
  method: 'GET',
  headers: {
    Accept: 'application/vnd.api+json',
  }
};

export const getRoutes = () => fetch(`${MTA_API_URL}/routes?filter%5Btype%5D=0%2C1`, options);
export const getStops = (routeId) => fetch(`${MTA_API_URL}/stops?filter%5Broute%5D=${routeId}`, options);
export const getPredictedSchedule = (directionId, routeId, stopId) => fetch(`${MTA_API_URL}/predictions?filter%5Bdirection_id%5D=${directionId}&filter%5Broute%5D=${routeId}&filter%5Bstop%5D=${stopId}`)

export const makeRequest = async (request, ...args) => {
  console.log('args:', ...args);
  return new Promise(function(resolve, reject) {
    request(...args)
    .then(async (response)=>{
      console.log('response:', response);
      if (response && response.status === 200){
        let responseJson = response.json();
        resolve(responseJson);
        return;
      } else {
        reject(response.error)
        return;
      }
    })
    .catch((e) => {
      reject(e)
    })
  })
}
