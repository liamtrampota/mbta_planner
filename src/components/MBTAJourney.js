import React from 'react';
import '../styles/MTAJourney.css';
import * as api from '../api';
import StepContainer from './StepContainer';
import DepartureStepContainer from './steps/DepartureStepContainer';
import DirectionStepContainer from './steps/DirectionStepContainer';
import RouteStepContainer from './steps/RouteStepContainer';
import StopStepContainer from './steps/StopStepContainer';

export default class MBTAJourney extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      step: 0,
      routes: null,
      stops: {},
      selectedStepData: [null, null, null],
      predictedDepartureSchedule: [],
    }
  }

  componentWillMount = () => {
    this.getRoutes();
  }

  getRoutes = () => {
    api.makeRequest(api.getRoutes)
    .then((response) => {
      this.setState({ routes: response.data })
    })
    .catch((e) => {
      console.log('getRoutes e:', e);
    })
  }

  getStops = () => {
    const { stops, selectedStepData } = this.state;
    const selectedRouteId = selectedStepData[0].id;
    if(!stops[selectedRouteId]) {
      api.makeRequest(api.getStops, selectedRouteId)
      .then((response) => {
        stops[selectedRouteId] = response.data;
        this.setState({ stops })
      })
      .catch((e) => {
        console.log('getStops e:', e);
      })
    }
  }

  getDepartureTime = () => {
    const { selectedStepData } = this.state;
    const selectedRouteId = selectedStepData[0].id;
    const selectedStopId = selectedStepData[1].id;
    const selectedDirectionId = selectedStepData[2].directionId;
    api.makeRequest(api.getPredictedSchedule, selectedDirectionId, selectedRouteId, selectedStopId)
    .then((response) => {
      this.setState({ predictedDepartureSchedule: response.data })
    })
    .catch((e) => {
      console.log('getDepartureTime e:', e);
    })
  }

  selectData = (data) => {
    const updatedSelectedStepData = this.state.selectedStepData.slice();
    updatedSelectedStepData[this.state.step] = data;
    this.setState({ selectedStepData: updatedSelectedStepData });
  }

  goToNextStep = () => {
    if(this.state.step === 0){
      this.getStops();
    }
    if(this.state.step === 2){
      this.getDepartureTime();
    }
    this.setState({ step: this.state.step+1 });
  }

  goToPreviousStep = () => {
    const { selectedStepData, step } = this.state;
    if(step === 3){
      this.setState({ predictedDepartureSchedule: [], step: step-1 })
      return;
    }
    const updatedSelectedStepData = selectedStepData.slice();
    updatedSelectedStepData[step] = null;
    this.setState({ step: step-1, selectedStepData: updatedSelectedStepData });
  }

  render() {
    const { step, routes, stops, selectedStepData, predictedDepartureSchedule } = this.state;
    const selectedRoute = selectedStepData[0];
    const selectedStop = selectedStepData[1];
    const selectedDirection = selectedStepData[2];
    return (
      <div className='container'>
        <h1>Numerated MBTA Trip Planner</h1>
        <div>
          <h2>
            <span style={{color: step >= 0 ? selectedRoute ? 'green' : 'orange' : 'black'}}>Route</span>
            <span> -> </span>
            <span style={{color: step >= 1 ? selectedStop ? 'green' : 'orange' : 'black'}}>Stop</span>
            <span> -> </span>
            <span style={{color: step >= 2 ? selectedDirection ? 'green' : 'orange' : 'black'}}>Direction</span>
            <span> -> </span>
            <span style={{color: step >= 3 ? 'green' : 'black'}}>Schedule</span>
          </h2>
        </div>
        { step === 0 &&
          <StepContainer id="step0" stepIndex={0} nextStepEnabled={selectedRoute} goToNextStep={this.goToNextStep} goToPreviousStep={this.goToPreviousStep}>
            <RouteStepContainer
              id="routeStep"
              routes={routes}
              selectRoute={this.selectData}
              selectedRoute={selectedRoute}
            />
          </StepContainer>
        }
        { step === 1 &&
          <StepContainer stepIndex={1} nextStepEnabled={selectedStop} goToNextStep={this.goToNextStep} goToPreviousStep={this.goToPreviousStep}>
            <StopStepContainer
              stops={stops[selectedRoute.id]}
              selectStop={this.selectData}
              selectedStop={selectedStop}
            />
          </StepContainer>
        }
        { step === 2 &&
          <StepContainer id="step2" stepIndex={2} nextStepEnabled={selectedDirection} goToNextStep={this.goToNextStep} goToPreviousStep={this.goToPreviousStep}>
            <DirectionStepContainer
              selectDirection={this.selectData}
              selectedRoute={selectedRoute}
              selectedDirection={selectedDirection}
            />
          </StepContainer>
        }
        { step === 3 &&
          <StepContainer stepIndex={3} goToNextStep={this.goToNextStep} goToPreviousStep={this.goToPreviousStep}>
            <DepartureStepContainer
              predictedDepartureSchedule={predictedDepartureSchedule}
            />
          </StepContainer>
        }
      </div>
    )
  }
}
