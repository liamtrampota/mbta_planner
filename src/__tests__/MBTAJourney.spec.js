import React from 'react';
import { shallow } from 'enzyme';
import MBTAJourney from '../components/MBTAJourney';
import * as api from '../api';

jest.mock('../api', () => ({
  getRoutes: jest.fn(),
  getStops: jest.fn(),
  getPredictedSchedule: jest.fn(),
  makeRequest: jest.fn(),
}))

beforeEach(() => {
  jest.restoreAllMocks();
})

describe('MBTAJourney', () => {
  test('renders MBTAJourney', () => {
    api.makeRequest.mockImplementationOnce(() => Promise.resolve({ data: [] }));
    const wrapper = shallow(<MBTAJourney/>);
    expect(wrapper).toMatchSnapshot();
  });
  test('componentWillMount', async () => {
    api.makeRequest.mockImplementationOnce(() => Promise.resolve({ data: 'routes' }));
    const makeRequestSpy = jest.spyOn(api, 'makeRequest');
    const wrapper = await shallow(<MBTAJourney/>);
    expect(makeRequestSpy).toHaveBeenCalledTimes(1);
    expect(makeRequestSpy).toHaveBeenCalledWith(api.getRoutes);
    wrapper.update();
    expect(wrapper.state().routes).toEqual('routes');
  });
  describe('selectData', () => {
    it('updates state with data at the step index', async () => {
      const mockSelectedRoute = { id: 'route1' };
      api.makeRequest.mockImplementationOnce(() => Promise.resolve({ data: 'stops' }));
      const makeRequestSpy = jest.spyOn(api, 'makeRequest');
      const wrapper = await shallow(<MBTAJourney/>);
      wrapper.find('#routeStep').invoke('selectRoute')(mockSelectedRoute);
      expect(wrapper.state().selectedStepData[0]).toEqual(mockSelectedRoute);
    });
  });
  describe('goToNextStep', () => {
    it('when step is 0 gets stops', async () => {
      const mockSelectedRoute = { id: 'route1' };
      const expectedStateStops = { route1: 'stops' };
      api.makeRequest.mockImplementation(() => Promise.resolve({ data: 'stops' }));
      const makeRequestSpy = jest.spyOn(api, 'makeRequest');
      const wrapper = await shallow(<MBTAJourney/>);
      wrapper.find('#routeStep').invoke('selectRoute')(mockSelectedRoute);
      await wrapper.find('#step0').props().goToNextStep();
      wrapper.update();
      expect(wrapper.state().stops).toEqual(expectedStateStops);
    });
    it('when step is 2 gets departure times', async () => {
      const mockSelectedRoute = { id: 'route1' };
      api.makeRequest.mockImplementation(() => Promise.resolve({ data: 'departureTimes' }));
      const makeRequestSpy = jest.spyOn(api, 'makeRequest');
      const wrapper = await shallow(<MBTAJourney/>);
      wrapper.setState({ step: 2, selectedStepData: [{ id: 'route1'}, {id: 'stop1'}, {directionId: 'direction1'}]})
      wrapper.update();
      await wrapper.find('#step2').props().goToNextStep();
      wrapper.update();
      expect(wrapper.state().predictedDepartureSchedule).toEqual('departureTimes');
    });
  });
});
