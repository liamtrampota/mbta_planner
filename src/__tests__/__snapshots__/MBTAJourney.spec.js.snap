// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MBTAJourney renders MBTAJourney 1`] = `ShallowWrapper {}`;
